File: frontend/src/researchInterviewContracts.ts
import { z } from "zod";

import { getJson, postJson } from "./apiClient";
import { sha256DigestSchema } from "./mediaContracts";

const identifierSchema = z.string().uuid();
const timestampSchema = z.string().datetime({ offset: true });

export const researchInterviewCitationSchema = z.object({
  position: z.number().int().min(0).max(19),
  source_kind: z.literal("research_run"),
  target_id: identifierSchema,
  source_label: z.string().trim().min(1).max(500),
  quote: z.string().min(1).max(500),
  start_offset: z.number().int().min(0),
  end_offset: z.number().int().min(1),
}).strict().refine(
  (item) => item.end_offset - item.start_offset === item.quote.length,
  { message: "Research interview citation offsets must span the quote" },
);

const researchInterviewPersonaSchema = z.object({
  id: identifierSchema,
  position: z.number().int().min(0).max(7),
  persona_id: z.string().min(1).max(128),
  display_name: z.string().trim().min(1).max(200),
  profile_sha256: sha256DigestSchema,
}).strict();

export const researchPersonaInterviewSchema = z.object({
  id: identifierSchema,
  research_project_id: identifierSchema,
  research_simulation_run_id: identifierSchema,
  run_spec_sha256: sha256DigestSchema,
  graph_memory_sha256: sha256DigestSchema,
  cohort_id: identifierSchema,
  cohort_sha256: sha256DigestSchema,
  persona: researchInterviewPersonaSchema,
  question: z.string().trim().min(2).max(1000),
  source_sha256: sha256DigestSchema,
  interview_sha256: sha256DigestSchema,
  model_name: z.string().trim().min(1).max(200),
  semantic_config_sha256: sha256DigestSchema,
  prompt_schema_version: z.literal("sandowl-run-persona-interview/v1"),
  status: z.enum(["queued", "running", "succeeded", "failed"]),
  created_at: timestampSchema,
  started_at: timestampSchema.nullable(),
  completed_at: timestampSchema.nullable(),
  answer_markdown: z.string().min(1).max(2000).nullable(),
  citations: z.array(researchInterviewCitationSchema).max(20),
  answer_sha256: sha256DigestSchema.nullable(),
  error_code: z.string().min(1).max(128).nullable(),
  error_message: z.string().min(1).max(500).nullable(),
}).strict().superRefine((item, context) => {
  const terminal = item.status === "succeeded" || item.status === "failed";
  const timestampsValid = item.status === "queued"
    ? item.started_at === null && item.completed_at === null
    : item.started_at !== null && (terminal ? item.completed_at !== null : item.completed_at === null);
  const outputValid = item.status === "succeeded"
    ? item.answer_markdown !== null && item.citations.length > 0 && item.answer_sha256 !== null
      && item.error_code === null && item.error_message === null
    : item.answer_markdown === null && item.citations.length === 0 && item.answer_sha256 === null
      && (item.status === "failed"
        ? item.error_code !== null && item.error_message !== null
        : item.error_code === null && item.error_message === null);
  if (!timestampsValid || !outputValid) {
    context.addIssue({ code: z.ZodIssueCode.custom, message: "Research interview lifecycle is invalid" });
  }
  if (!item.citations.every((citation, position) => citation.position === position)) {
    context.addIssue({ code: z.ZodIssueCode.custom, message: "Research interview citations are not contiguous" });
  }
});

export const researchPersonaInterviewsResponseSchema = z.object({
  items: z.array(researchPersonaInterviewSchema),
  total: z.number().int().min(0),
}).strict().refine((response) => response.total === response.items.length, {
  message: "Research interview total must match items",
});

export const researchPersonaInterviewSessionSchema = z.object({
  id: identifierSchema,
  research_project_id: identifierSchema,
  research_simulation_run_id: identifierSchema,
  run_spec_sha256: sha256DigestSchema,
  graph_memory_sha256: sha256DigestSchema,
  cohort_id: identifierSchema,
  cohort_sha256: sha256DigestSchema,
  question: z.string().trim().min(2).max(1000),
  persona_count: z.number().int().min(2).max(8),
  session_sha256: sha256DigestSchema,
  model_name: z.string().trim().min(1).max(200),
  semantic_config_sha256: sha256DigestSchema,
  prompt_schema_version: z.literal("sandowl-run-persona-interview-session/v1"),
  status: z.enum(["queued", "running", "succeeded", "failed"]),
  created_at: timestampSchema,
  interviews: z.array(researchPersonaInterviewSchema).min(2).max(8),
}).strict();

export type ResearchPersonaInterview = z.infer<typeof researchPersonaInterviewSchema>;
export type ResearchPersonaInterviewsResponse = z.infer<typeof researchPersonaInterviewsResponseSchema>;
export type ResearchPersonaInterviewSession = z.infer<typeof researchPersonaInterviewSessionSchema>;

function runPath(projectId: string, runId: string): string {
  return `/api/v2/research-projects/${encodeURIComponent(identifierSchema.parse(projectId))}/runs/${encodeURIComponent(identifierSchema.parse(runId))}`;
}

export function fetchResearchPersonaInterviews(
  projectId: string,
  runId: string,
  signal: AbortSignal,
): Promise<ResearchPersonaInterviewsResponse> {
  return getJson(
    `${runPath(projectId, runId)}/persona-interviews`,
    researchPersonaInterviewsResponseSchema,
    signal,
  );
}

export function createResearchPersonaInterview(
  projectId: string,
  runId: string,
  personaId: string,
  question: string,
  signal: AbortSignal,
): Promise<ResearchPersonaInterview> {
  return postJson(
    `${runPath(projectId, runId)}/persona-interviews`,
    { persona_id: identifierSchema.parse(personaId), question },
    researchPersonaInterviewSchema,
    signal,
  );
}

export function createResearchPersonaInterviewSession(
  projectId: string,
  runId: string,
  personaIds: readonly string[],
  question: string,
  signal: AbortSignal,
): Promise<ResearchPersonaInterviewSession> {
  return postJson(
    `${runPath(projectId, runId)}/persona-interview-sessions`,
    { persona_ids: personaIds.map((item) => identifierSchema.parse(item)), question },
    researchPersonaInterviewSessionSchema,
    signal,
  );
}

File: backend/oasis_worker/src/oasis_worker/research_interview_contracts.py
"""Strict worker contracts for interviews over a frozen research run."""

from datetime import datetime
from typing import Annotated, Literal
from uuid import UUID

from pydantic import Field, StringConstraints, field_validator

from oasis_worker.contracts import RequiredText, Sha256, StrictModel
from oasis_worker.semantic_contracts import PersonaProfile


class ClaimedResearchPersonaInterview(StrictModel):
    id: UUID
    research_project_id: UUID
    research_simulation_run_id: UUID
    run_spec_sha256: Sha256
    graph_memory_sha256: Sha256
    cohort_id: UUID
    cohort_sha256: Sha256
    persona_id: UUID
    persona_position: Annotated[int, Field(ge=0, le=7)]
    persona_external_id: Annotated[RequiredText, Field(max_length=128)]
    persona_display_name: Annotated[RequiredText, Field(max_length=200)]
    persona_profile: PersonaProfile
    persona_profile_sha256: Sha256
    question: Annotated[str, StringConstraints(min_length=2, max_length=1000, strict=True)]
    source_text: Annotated[str, StringConstraints(min_length=1, max_length=80000, strict=True)]
    source_sha256: Sha256
    interview_sha256: Sha256
    model_name: Annotated[RequiredText, Field(max_length=200)]
    semantic_config_sha256: Sha256
    prompt_schema_version: Literal["sandowl-run-persona-interview/v1"]
    created_at: datetime


class ExtractedResearchPersonaInterviewAnswer(StrictModel):
    answer_markdown: Annotated[
        str,
        StringConstraints(min_length=1, max_length=2000, strict=True),
        Field(description="A bounded first-person synthetic perspective."),
    ]
    citation_quotes: Annotated[
        tuple[Annotated[str, StringConstraints(min_length=1, max_length=500, strict=True)], ...],
        Field(min_length=1, max_length=20),
    ]

    @field_validator("citation_quotes")
    @classmethod
    def require_unique_quotes(cls, value: tuple[str, ...]) -> tuple[str, ...]:
        if len(value) != len(set(value)):
            raise ValueError("citation_quotes must be unique")
        return value


class ResearchInterviewCitation(StrictModel):
    position: Annotated[int, Field(ge=0, le=19)]
    source_kind: Literal["research_run"]
    target_id: UUID
    source_label: Annotated[RequiredText, Field(max_length=500)]
    quote: Annotated[str, StringConstraints(min_length=1, max_length=500, strict=True)]
    start_offset: Annotated[int, Field(ge=0)]
    end_offset: Annotated[int, Field(ge=1)]


class NormalizedResearchPersonaInterviewAnswer(StrictModel):
    answer_markdown: Annotated[str, StringConstraints(min_length=1, max_length=2000, strict=True)]
    citations: Annotated[tuple[ResearchInterviewCitation, ...], Field(min_length=1, max_length=20)]
    answer_sha256: Sha256
